rtmr/extend/extend.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extend
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/sha512" // Registrer SHA384 and SHA512
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/google/go-configfs-tsm/configfs/configfsi"
+	"github.com/google/go-configfs-tsm/configfs/linuxtsm"
+	"github.com/google/logger"
+)
+
+const (
+	RtmrSubsystem = "rtmrs"
+	TsmRtmrPrefix = configfsi.TsmPrefix + "/" + RtmrSubsystem
+	TsmRtmrDigest = "digest"
+	TsmPathIndex  = "index"
+	TsmPathTcgMap = "tcg_map"
+)
+
+// RtmrExtend is a struct that represents a rtmr entry in the configfs.
+type RtmrExtend struct {
+	RtmrIndex int
+	entry     *configfsi.TsmPath
+	client    configfsi.Client
+}
+
+// Remove the null bytes from the c-string
+func convertCstring(b []byte) string {
+	return string(bytes.Trim(b, "\x00"))
+}
+
+func (r *RtmrExtend) attribute(subtree string) string {
+	a := *r.entry
+	a.Attribute = subtree
+	return a.String()
+}
+
+// ExtendDigest extends the measurement to the rtmr with the given hash.
+func (r *RtmrExtend) ExtendDigest(hash []byte) error {
+	if err := r.client.WriteFile(r.attribute(TsmRtmrDigest), hash); err != nil {
+		if (r.RtmrIndex == 2) || (r.RtmrIndex == 3) {
+			return fmt.Errorf("could not write digest to rmtr%d: %v. Is your rtmr index extendable from userspace?", r.RtmrIndex, err)
+		}
+		return fmt.Errorf("could not write digest to rmtr%d: %v", r.RtmrIndex, err)
+	}
+	return nil
+}
+
+// SetRtmrIndex sets a configfs rtmr entry to the given index.
+// It reports an error if the index cannot be written or the rtmr_tcg map does not match.
+func (r *RtmrExtend) SetRtmrIndex() error {
+	indexBytes := []byte(strconv.Itoa(r.RtmrIndex)) // Convert index to []byte
+	indexPath := r.attribute(TsmPathIndex)
+	if err := r.client.WriteFile(indexPath, indexBytes); err != nil {
+		return fmt.Errorf("could not write index %s: %v", indexPath, err)
+	}
+	if err := r.ValidateRtmrIndex(); err != nil {
+		return fmt.Errorf("the tcg_map is invalid %s: %v", indexPath, err)
+	}
+	return nil
+}
+
+// ValidateRtmrIndex checks if the rtmr to PCR map is valid.
+// It returns an error if the rtmr_tcg map does not match the expected value.
+func (r *RtmrExtend) ValidateRtmrIndex() error {
+	if (r.client == nil) || (r.entry == nil) {
+		return fmt.Errorf("RtmrExtend is not initialized")
+	}
+	data, err := r.client.ReadFile(r.attribute(TsmPathTcgMap))
```

Comment:
Every time when extending one rtmr entry, the config will be read and processed again, this seems little bit inefficient, is this by design?

Reply:
Yeah, it is by design. The mapping config file is very small and we need to make sure we write the digest into the correct entry.